### ISSUE TITLE
Issues computing RDC

### ISSUE THREAD

**Issue by ivan-marroquin** [garydoranjr/rdc#1]
To  whom it may concern,

I have Python 3.6.5 on a windows machine. Downloaded the rdc.py file, and imported into my Python script. To test the correlation measure, 2d synthetic data sets were generated (http://minepy.sourceforge.net/docs/1.0.0/python.html).

When I run my script, these are the encountered problems:
A) slice indices must be integers or None or have an __index__ method

This is caused at line with command: Cxx = C[:k, :k]. I noticed that the k value becomes a real number at the second iteration of the while loop.

To fix this issue, i added this comman line: k= np.round(k).astype(np.int)

B) LinAlgError: Singular matrix

This is caused at line with command:  eigs = np.linalg.eigvals(np.dot(np.dot(np.linalg.inv(Cxx), Cxy),  np.dot(np.linalg.inv(Cyy), Cyx)))

for this issue, I couldn't find a way to fix it. 

I attached a copy of my Python script and input data.

Many thanks,
Ivan
[Python_RDC_2D_Synthetic_Data.zip](https://github.com/garydoranjr/rdc/files/3131365/Python_RDC_2D_Synthetic_Data.zip)

**Comment by garydoranjr** [comment]
@ivan-marroquin,

Thanks for letting me know about this issue. This code was originally written for Python 2, so the issue with the index becoming a floating point number is caused by differences in how Python 3 handles integer division. I've updated the code to force it to use integer division.

I wasn't able to reproduce the singular matrix error. When I run your script, the code terminates successfully with the following output:

    Python Version:
    3.6.6 (default, Oct  2 2018, 10:03:30) 
    [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.2)]
    RDC correlation measures on 2D synthetic linear data
    
    [[1.   ]
     [0.784]
     [0.445]
     [0.067]
     [0.366]
     [0.797]
     [1.   ]]
    
    
    RDC correlation measures on 2D synthetic nonlinear data
    
    [[0.475]
     [0.262]
     [0.29 ]
     [0.859]
     [0.731]
     [0.914]
     [0.026]]

I'm using numpy version 1.15.2 and scipy version 1.1.0.

**Comment by ivan-marroquin** [comment]
Hi @garydoranjr 

Many thanks for updating the script to Python 3. I download it and run it with my script. Unfortunately, I still get the same error message:

Traceback (most recent call last):
  File "Python_RDC_2D_Synthetic_Data.py", line 37, in <module>
    Statistics[count,0]= rdc(pairs_linear_data[:,i].reshape(-1), pairs_linear_data[:,(i+1)].reshape(-1))
  File "C:\Users\IMarroquin\Documents\Publish_Documentation\2019\Independence_Measures\Python_scripts_data\RDC_python_code\rdc.py", line 77, in rdc
    np.dot(np.linalg.inv(Cyy), Cyx)))
  File "C:\Temp\Python\Python36\lib\site-packages\numpy\linalg\linalg.py", line 532, in inv
    ainv = _umath_linalg.inv(a, signature=signature, extobj=extobj)
  File "C:\Temp\Python\Python36\lib\site-packages\numpy\linalg\linalg.py", line 89, in _raise_linalgerror_singular
    raise LinAlgError("Singular matrix")
numpy.linalg.linalg.LinAlgError: Singular matrix


My installation has numpy 1.15.4 and scipy 1.1.0

I have practically the same package version as you. Is there a way to test if my installation is corrupted?

Once again, thanks a lot!

Ivan

**Comment by ivan-marroquin** [comment]
I just removed numpy to install version 1.15.2, and I still got the same issue

**Comment by garydoranjr** [comment]
Hi @ivan-marroquin,

Thanks for trying again with different numpy versions. I ran the code on a Windows virtual machine and was able to reproduce the issue. My best guess is that the linear algebra library matrix inversion is implemented slightly differently for Windows and causes numerical issues to result. I changed the the inverse to a pseudo inverse, which should be more numerically stable. When I do this, I am able to run the code on Windows and produce the same result as shown above. I pushed a change that implements this, so let me know if this works for you, too.

-Gary
